### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by kelvin8773**
Finally, finish another milestone, try to learn more about React testing.
* Test for App (modules load and accessibility test)
* test for header component
* test for footer component
* test for main index page
* test for search functions (search, heatmap, portable)

### CODE REVIEW

File: src/components/navbar/index.test.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+const Navlinks = [
+  { text: 'text-to-not-found', location: '/' },
+  { text: 'Search', location: '/search/javascript' },
+  { text: 'How it works', location: '/#howItWorks' },
+  { text: 'About', location: '/#about' },
+];
+
+describe('NavBar links', () => {
+  test.each(Navlinks)(
```
Comment (jkettmann) [inline src/components/navbar/index.test.js:16]: Here also write 4 separate tests. Or one for the logo and one with iteration for the other links

File: src/components/home/infoSection.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ const InfoLine = Styled.div`
 
 const InfoSection = () => (
   <Section>
-    <Info id="howItWorks">
+    <Info id="howItWorks" data-testid="howItWorks">
```
Comment (jkettmann) [inline src/components/home/infoSection.js:36]: test id probably not necessary

File: src/components/home/infoSection.js [hunk 3]
```diff
@@ -49,7 +49,7 @@ const InfoSection = () => (
       </InfoLine>
     </Info>
 
-    <Info id="about">
+    <Info id="about" data-testid="about">
```
Comment (jkettmann) [inline src/components/home/infoSection.js:52]: test id probably not necessary

File: src/components/navbar/index.js [hunk 4]
```diff
@@ -22,19 +23,18 @@ const LinkItems = Styled.div`
...
 `;
 
-
 const Navbar = () => {
   const subReddit = SearchJson.defaultSubreddit;
 
   return (
-    <StyledNavbar>
-      <Link to="/">
+    <StyledNavbar data-testid="header">
```
Comment (jkettmann) [inline src/components/navbar/index.js:36]: test id probably not necessary

File: src/components/navbar/index.js [hunk 5]
```diff
@@ -22,19 +23,18 @@ const LinkItems = Styled.div`
...
+      <Link to="/" data-testid="navLogo">
```
Comment (jkettmann) [inline src/components/navbar/index.js:37]: test id probably not necessary

File: src/components/search/postTable.js [hunk 6]
```diff
@@ -60,7 +60,7 @@ const PostTable = ({ posts }) => {
   const sortedPosts = [...posts].sort((a, b) => getMin(a.created_utc) - getMin(b.created_utc));
 
   return (
-    <div>
+    <div data-testid="postTable">
```
Comment (jkettmann) [inline src/components/search/postTable.js:63]: test id probably not necessary

File: src/components/search/searchForm.js [hunk 7]
```diff
@@ -40,35 +40,50 @@ const SearchInput = Styled.input`
...
+    }
   };
 
   useEffect(() => {
     setSubreddit(redditName);
   }, [redditName]);
 
   return (
-    <Container>
+    <Container data-testid="searchForm">
```
Comment (jkettmann) [inline src/components/search/searchForm.js:72]: test id probably not necessary

File: src/components/search/spinner.js [hunk 8]
```diff
@@ -11,9 +12,16 @@ const rotate = keyframes`
...
 `;
 
-const Spinner = Styled(spinner)`
+const StyledSpinner = Styled(spinner)`
   margin-top: 56px;
   animation: ${rotate} 1.5s linear infinite;
 `;
 
+const Spinner = () => (
+  <div data-testid="loadSpinner">
```
Comment (jkettmann) [inline src/components/search/spinner.js:21]: test id might not be necessary

File: src/helper/redditAPI.js [hunk 9]
```diff
@@ -1,6 +1,9 @@
 import axios from 'axios';
 import dayjs from 'dayjs';
 
+// eslint-disable-next-line no-console
+console.log('use real API');
+
```
Comment (jkettmann) [inline src/helper/redditAPI.js:6]: should be removed

File: src/helper/redditAPI.js [hunk 10]
```diff
@@ -13,13 +16,24 @@ const getPosts = async (subreddit) => {
...
-      return response.data.data;
+    if (response.status === 200) {
+      return response.data.data.reduce((result, post) => {
+        result.push({
+          id: post.id,
+          title: post.title,
+          full_link: post.full_link,
+          created_utc: post.created_utc,
+          score: post.score,
+          num_comments: post.num_comments,
```
Comment (jkettmann) [inline src/helper/redditAPI.js:27]: At this place you could actually rename the fields to fit the general naming rules. For example, `numComments` instead of `num_comments`

File: src/helper/saveMockPosts.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
+// Run with Node.js
```
Comment (jkettmann) [inline src/helper/saveMockPosts.js:1]: It's great to have scripts like this. A typical location might be `/scripts` outside the `/src` folder. This way it's clear that this is not used by the application. 

Not sure if you did that already but you could add a script in `package.json` that runs this code

File: src/pages/index.js [hunk 12]
```diff
@@ -14,12 +14,11 @@ const StyledMain = Styled.div`
 `;
 
 const Main = () => (
-  <StyledMain>
+  <StyledMain data-testid="main">
```
Comment (jkettmann) [inline src/pages/index.js:17]: test id probably not necessary

File: src/pages/index.test.js [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+  });
+
+  test('click Hero Img to enter search Page', async () => {
+    const history = createMemoryHistory();
+    setup(<Main />, history);
+    const heroImg = screen.getByAltText(/heat map/i);
+    expect(heroImg).toBeInTheDocument();
+    user.click(heroImg);
+
+    expect(screen.queryByText(/search/i)).toBeInTheDocument();
```
Comment (jkettmann) [inline src/pages/index.test.js:49]: use `getByRole` instead

File: src/pages/index.test.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+import { render, screen, act } from '@testing-library/react';
+import user from '@testing-library/user-event';
```
Comment (jkettmann) [inline src/pages/index.test.js:3]: I'd name this `userEvent` since most developer will be used to that name

File: src/pages/index.test.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    setup(<Main />, history);
+    const heroImg = screen.getByAltText(/heat map/i);
+    expect(heroImg).toBeInTheDocument();
+    user.click(heroImg);
+
+    expect(screen.queryByText(/search/i)).toBeInTheDocument();
+    expect(screen.queryByDisplayValue(searchJson.defaultSubreddit)).toBeInTheDocument();
+    expect(history.location.pathname).toEqual(`/search/${searchJson.defaultSubreddit}`);
+
+    await act(() => Promise.resolve());
```
Comment (jkettmann) [inline src/pages/index.test.js:53]: Why is this required? I guess you can remove it

File: src/pages/index.test.js (hunk 13 above)
Comment (jkettmann) [reply src/pages/index.test.js:49]: Is this searching for the subreddit form button? Because I think it's testing the header search link

File: src/pages/index.test.js [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+  test('click Hero Img to enter search Page', async () => {
+    const history = createMemoryHistory();
+    setup(<Main />, history);
+    const heroImg = screen.getByAltText(/heat map/i);
+    expect(heroImg).toBeInTheDocument();
+    user.click(heroImg);
+
+    expect(screen.queryByText(/search/i)).toBeInTheDocument();
+    expect(screen.queryByDisplayValue(searchJson.defaultSubreddit)).toBeInTheDocument();
```
Comment (jkettmann) [inline src/pages/index.test.js:50]: If you're trying to check that the correct page is rendered you can probably just search for the form with `getByRole` instead of this and the previous assertion

File: src/pages/index.test.js [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    expect(history.location.pathname).toEqual(`/search/${searchJson.defaultSubreddit}`);
+
+    await act(() => Promise.resolve());
+  });
+
+  test('click Home Button to enter search Page', async () => {
+    const history = createMemoryHistory();
+    setup(<Main />, history);
+    const homeButton = screen.getByText(/show me the best time/i);
```
Comment (jkettmann) [inline src/pages/index.test.js:59]: rather use getByRole

File: src/pages/index.test.js [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+    expect(homeButton).toBeInTheDocument();
```
Comment (jkettmann) [inline src/pages/index.test.js:60]: Not sure if this is required. If it's not in the document the test will fail anyway. I guess I'd remove this line

File: src/pages/index.test.js [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+    user.click(homeButton);
+
+    expect(screen.queryByText(/search/i)).toBeInTheDocument();
+    expect(screen.queryByDisplayValue(searchJson.defaultSubreddit)).toBeInTheDocument();
```
Comment (jkettmann) [inline src/pages/index.test.js:64]: Same as in the previous test: You might want to check if the form is there

File: src/pages/index.test.js [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    expect(history.location.pathname).toEqual(`/search/${searchJson.defaultSubreddit}`);
+
+    await act(() => Promise.resolve());
```
Comment (jkettmann) [inline src/pages/index.test.js:67]: Not required I guess

File: src/pages/index.test.js [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+  });
+
+
+  const routes = [
+    { path: '/', showText: 'r/javascript' },
+    { path: '/search/javascript', showText: 'javascript' },
+    { path: '/search/reactjs', showText: 'reactjs' },
+  ];
+
+  test.each(routes)('render different %s route', async (route) => {
```
Comment (jkettmann) [inline src/pages/index.test.js:77]: I'd separate this into multiple tests. You also basically tested the routes in the previous tests already. So maybe you can remove this test

File: src/helper/saveMockPosts.js (hunk 11 above)
Comment (kelvin8773) [reply src/helper/saveMockPosts.js:1]: that is a great idea!

File: src/App.test.js [hunk 22]
```diff
@@ -1,9 +1,32 @@
...
-  const linkElement = getByText(/javascript/i);
-  expect(linkElement).toBeInTheDocument();
+jest.mock('./helper/redditAPI');
+
+expect.extend(toHaveNoViolations);
+
+describe('App', () => {
+  test('header should load', () => {
+    render(<App />);
+    expect(screen.getByTestId('header')).toBeInTheDocument();
```
[the remark this answers is not in the document]
Comment (kelvin8773) [reply src/App.test.js:13]: right, let's use roles instead.

File: src/components/footer/index.test.js [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+import Footer from './index';
+
+const Links = [
+  { text: 'ooloo.io', location: 'https://ooloo.io' },
+  { text: 'link for home', location: '/' },
+  { text: 'Terms & Privacy', location: '/terms' },
+];
+
+describe('Footer Links', () => {
+  test.each(Links)(
```
[the remark this answers is not in the document]
Comment (kelvin8773) [reply src/components/footer/index.test.js:13]: but I think I think I prefer to keep it since it is more flexible in different situations.

File: src/pages/index.test.js (hunk 13 above)
Comment (kelvin8773) [reply src/pages/index.test.js:49]: role is not working here, since home page also have button. I use tested instead.

File: src/pages/index.test.js (hunk 14 above)
Comment (kelvin8773) [reply src/pages/index.test.js:3]: ok, fair. But I do think use 'user' are more readable in the code.
`userEvent.click(homeButton);`
vs
`user.click(homeButton);`

File: src/pages/index.test.js (hunk 15 above)
Comment (kelvin8773) [reply src/pages/index.test.js:53]: it is for Fix the "not wrapped in act(...)" warning, you can check out the detail [here](https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning).

File: src/pages/index.test.js (hunk 17 above)
Comment (kelvin8773) [reply src/pages/index.test.js:59]: same reason, another page also have button, I will keep this one.

File: src/pages/index.test.js (hunk 21 above)
Comment (kelvin8773) [reply src/pages/index.test.js:75]: I would keep it for now, it is nothing wrong to double test from different methods.

File: src/components/footer/index.test.js [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+    'Check if have %s link.',
+    (link) => {
+      render(
+        <BrowserRouter>
+          <Footer />
+        </BrowserRouter>,
+      );
+      const linkDom = screen.queryByText(link.text) || screen.getByTestId('footLogo');
```
[the remark this answers is not in the document]
Comment (kelvin8773) [reply src/components/footer/index.test.js]: but I use as react component, can't see the file name.

File: src/components/navbar/index.test.js (hunk 1 above)
Comment (kelvin8773) [reply src/components/navbar/index.test.js:27]: separated with logo and links

File: src/components/navbar/index.js (hunk 4 above)
Comment (kelvin8773) [reply src/components/navbar/index.js]: removed.

File: src/components/home/infoSection.js (hunk 2 above)
Comment (kelvin8773) [reply src/components/home/infoSection.js:36]: what else I can use?

File: src/components/search/searchForm.js (hunk 7 above)
Comment (kelvin8773) [reply src/components/search/searchForm.js:72]: removed.

File: src/components/search/spinner.js (hunk 8 above)
Comment (kelvin8773) [reply src/components/search/spinner.js:21]: I think I need this one.

File: src/helper/redditAPI.js (hunk 9 above)
Comment (kelvin8773) [reply src/helper/redditAPI.js]: removed.

File: src/pages/index.js (hunk 12 above)
Comment (kelvin8773) [reply src/pages/index.js]: removed.

File: src/pages/index.test.js (hunk 16 above)
Comment (kelvin8773) [reply src/pages/index.test.js:50]: updated.

File: src/pages/index.test.js (hunk 18 above)
Comment (kelvin8773) [reply src/pages/index.test.js]: removed.

File: .vscode/settings.json [hunk 25]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (jkettmann) [reply .vscode/settings.json:1]: You can remove the file from tracking with `git rm --cached <file>`. Have a look [here](https://stackoverflow.com/questions/1274057/how-to-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore) for more details

File: src/components/footer/index.js [hunk 26]
```diff
@@ -25,17 +25,17 @@ const LinkItem = Styled(Link)`
 `;
 
 const Footer = () => (
-  <StyledFooter>
+  <StyledFooter role="contentinfo" aria-label="footer links">
```
Comment (jkettmann) [inline src/components/footer/index.js:28]: The `footer` element has the role `contentinfo` by default. So you can remove it here.

Out of curiosity: What is the `aria-label` used for?

File: src/components/footer/index.test.js [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+];
+
+describe('Footer Links', () => {
+  test('check if logo loaded', () => {
+    render(
+      <BrowserRouter>
+        <Footer />
+      </BrowserRouter>,
+    );
+    const logo = screen.getByTestId('footLogo');
```
Comment (jkettmann) [inline src/components/footer/index.test.js:18]: I tried to replace this line with `const logo = screen.getByRole('link', { name: /sign\.svg/ });` and it works. `sign.svg` is the filename of the svg where you import it from

Comment (jkettmann) [reply src/components/footer/index.test.js:18]: But you can keep it with test id if you like it better

File: src/components/home/infoSection.js (hunk 2 above)
Comment (jkettmann) [reply src/components/home/infoSection.js:36]: You could use an HTML `section` element for `Section` and an `article` element for `Info`. Then you can target them with `getByRole` again

File: src/pages/index.test.js (hunk 15 above)
Comment (jkettmann) [reply src/pages/index.test.js:53]: Interesting :smile:

File: src/pages/index.test.js (hunk 17 above)
Comment (jkettmann) [reply src/pages/index.test.js:59]: The other page is not so important here. If there was another button on the same page you might be right. But it's fine to keep it as is

File: src/components/footer/index.js (hunk 26 above)
Comment (kelvin8773) [reply src/components/footer/index.js:28]: Great tips!
Not very sure, just try to mimic [other roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Navigation_Role), but it seems not necessary now. let me remove it.

File: src/components/footer/index.test.js (hunk 27 above)
Comment (kelvin8773) [reply src/components/footer/index.test.js:18]: ok, let me try that, thanks again for your detailed advice.

Comment (kelvin8773) [reply src/components/footer/index.test.js:18]: yes, it work, I remove both footer and header logo testid, great!
